odoo/tools/html_sanitize.py

```diff
@@ -0,0 +1,294 @@
+
+    def sanitize_token(self, token):
+        tag_name = token.get('name', '').lower().strip()
+        tag_type = token['type']
+
+        if self.sanitize_form and tag_name == 'form':
+            # Ignore sanitize_tags, always remove it
+            return self._mark_to_kill(token)
+
+        if self.sanitize_tags and tag_name in kill_tags:
+            # We will strip manually those token because
+            # bleach will keep the child elements
+            return self._mark_to_kill(token)
+
+        if (not self.sanitize_tags
+           and tag_type in ('StartTag', 'EmptyTag', 'EndTag')
+           and tag_name not in clean_tags):
+            return self.allow_token(token)
+
+        # Sanitize attributes and tags
+        return super().sanitize_token(token)
+
+    def _mark_to_kill(self, token):
+        """Kill a token, the element and all its children will be removed.
+
+        We just keep the token type (StartTag, EndTag,...) and use a custom tag name
+        to be able to remove it after the bleach sanitation. This is needed because
+        bleach only remove the element itself and not its children.
+        """
+        return {
+            'name': KILLED_TAG_NAME,
+            'type': token['type'],
+            'data': {},
+        }
+
+    def sanitize_uri_value(self, value, allowed_protocols):
+        allowed_data_protocols = (
+            'data:image/png;base64,',
+            'data:image/jpg;base64,',
+            'data:image/jpeg;base64,',
+            'data:image/bmp;base64,',
+            'data:image/gif;base64,',
+        )
+
+        for protocol in allowed_data_protocols:
+            if value.startswith(protocol):
+                return value
+
+        return super().sanitize_uri_value(value, allowed_protocols)
+
+
+def html_sanitize(
+    src,
+    silent=True,
+    sanitize_tags=True,
+    sanitize_attributes=True,
+    sanitize_style=False,
+    sanitize_form=True,
+    strip_style=False,
+    strip_classes=False,
+):
+    """Sanitize an un-trusted HTML source to be safe from a XSS point of view.
+
+    Careful if you change one default value, it might create a XSS,
+    do not change them if you don't know what you do!
+
+    :param src: HTML string to sanitize
+    :param silent: If an error occurs during the parsing, do not raise
+    :param sanitize_tags: Allow only an allowed list of HTML tags
+    :param sanitize_attributes: Allow only an allowed list of attributes
+        If False, the allowed list will be extended
+    :param sanitize_style: Sanitize the style attribute, allow only an allowed list of CSS value
+    :param sanitize_form: Remove any <form/> tags
+    :param strip_style: Remove any "style" attribute, can not be used when sanitize_style is set
+    :param strip_classes: Remove any "class" attribute
+    """
+    if not src or not src.strip():
+        return src
+
+    if isinstance(src, bytes):
+        src = src.decode()
+
+    safe_tags_ = list(safe_tags)
```

Comment:
I know that bleach ask a `list` for `allowed_elements`, but if you look to the implementation : https://github.com/mozilla/bleach/blob/6cd4d527a6b43569c1e1490e632500199b1efb6c/bleach/sanitizer.py#L230 , `allowed_elements` is only use to make a `in` operator and the super Class (https://github.com/mozilla/bleach/blob/6cd4d527a6b43569c1e1490e632500199b1efb6c/bleach/_vendor/html5lib/filters/sanitizer.py#L724) ask a `set`, not a `list` (for efficiency I suppose).

Should you send the `safe_tags` directly (even if it doesn't respect their documentation - which is imcomplete/wrong) for performance ? Or I miss something ?

Reply:
good catch :)